---
title: "ATAC-seq analysis of SN4741 cells"
author: "Sarah McClymont"
date: "November 11, 2022"
output: pdf_document
---

# Intro 

Generated ATAC-seq in the SN4741 cell line at both the permissive and restrictive temperatures (n=4/temperature). Want to see how the chromatin differs between the two states. Also want to consider how well they recapitulate the ATAC-seq data from primary E15.5 mouse dopaminergic neurons from both the forebrain and midbrain, which are presumably the cell type this cell line should be modelling.

# Pre-alignment QC

Use FastQC to check library quality. Summarize FastQC reports with MultiQC and check the generated HTML for sequencing quality issues.

```{r, eval = F}

# batch script: fastqc.sh

#!/bin/bash
#SBATCH --job-name=fastqc
#SBATCH --time=8:00:00
#SBATCH -N 1
#SBATCH -n 10

module load fastqc 

## fastqc v0.11.9
fastqc -t 10 *.fastq.gz

# Generate report with MultiQC
## pip3 install --user multiqc #v1.13
multiqc .

# Clean up files
mkdir FastQC
mv *fastqc* FastQC

exit 0

```

MultiQC report moved to Output/multiqc_report_fastqc.html

**Conclusions:** Look good in general. Sample 39-4 looks off in its GC content compared to other samples. High rates of duplication in 37-1 and 37-4 but the pool was apparently unbalanced and we sequenced these two very deeply so perhaps hit saturation. Adaptor contamination for all samples is functionally zero - approx 0.25% of reads end with the beginning of an adaptor - not worth trimming and shouldn't affect mapping at all. As always, all samples have sequence composition bias at the start of reads, as per usual for ATAC. Nothing of concern here - move forward with mapping!


# Align and call peaks

Use Bowtie2 to align to genome.  
UCSC mm10 indexes are the same as those used in the pcHiC. See NBB's code for source.  
  Note: NBB's genome files don't contain chrM so can't quantify %mito. Previously aligned these sequences and saw less than 1% contamination in all samples.

Deduplicate with Picard Tools. 
Index with SAMtools. 
Call peaks with MACS3.

Summarize reports with MultiQC.   

```{r, eval = F}

# alignment.sh

#!/bin/bash
#SBATCH --job-name=alignment
#SBATCH --time=30:00:00
#SBATCH -N 1
#SBATCH -n 12

# Load modules
  module load bowtie2 #v2.4.1
  module load samtools #v1.15.1
  module load picard #v2.26.11

# Define path to genome indices
  genome_index=/data/amccall2/work-zfs/NBB/SN4741-pcHiC/bowtie2_index/mm10

for file in *L001_R1_001.fastq.gz; do

# Make life easier by making the file names simpler and into an extensionlesss variable
  sample=${file%_L001_R1_001.fastq.gz}

# Write out sample name so can track progress/errors in slurm.out file
  echo $sample

# Align with Bowtie2
  # Comma separate input files, no white space between
  # Local alignment to softclip the ends
  # Increase pair distance to 1000 to avoid the 500bp blip on fragment length plot
  # Write out alignment stats (stderr) to file for multiqc summary
  bowtie2 -p 12 --local -X 1000 -x $genome_index -1 $file,${sample}_L002_R1_001.fastq.gz -2 ${sample}_L001_R2_001.fastq.gz,${sample}_L002_R2_001.fastq.gz -S ${sample}.sam 2>${sample}.align.stats

# Convert to bam and sort with samtools
  # view -bu turns into uncompressed bam for easy piping into sort
  samtools view -bu ${sample}.sam | samtools sort -o ${sample}_sorted.bam

# Deduplicate with picard tools
  picard MarkDuplicates I=${sample}_sorted.bam O=${sample}.bam M=${sample}_markdup.stats

# Index with samtools
  samtools index ${sample}.bam

# Remove intermediate files
	${sample}.sam
	${sample}_sorted.bam

# Call peaks! macs3 v3.0.0a7
	macs3 callpeak --nomodel -B -f BAMPE -t ${sample}.bam -n $sample --nolambda --gsize mm

	done

exit 0
  
# Compile alignment statistics with multiqc
	multiqc . --ignore "FastQC/"
	
```

**Conclusions:** Libraries look great! High alignment rate (~98% for all eight samples), and average duplication rate (â‰¤20%). Two/three samples have high duplication rate, but those are also the deepest sequenced, so seems like we are hitting saturation for them. 

Arrange outputs into folders: alignment, bams, FastQC, fastqs, and peaks for easier navigation.

# Removing blacklisted regions

*Source of blacklists:*
ENCODE blacklist (requires citation to ENCODE): https://sites.google.com/site/anshulkundaje/projects/blacklists
ATAC-seq specific blacklists (hg19, mm9, mm10; must be member of ATAC-seq forum to view): https://sites.google.com/site/atacseqpublic/atac-seq-analysis-methods/mitochondrialblacklists-1

Concatenated the two mm10 blacklists together; remove peaks that overlap these regions.

```{r, eval = F}

# blacklisted.sh

#!/bin/bash
#SBATCH --job-name=blacklisted
#SBATCH --time=1:00:00

# Load modules
module load bedtools

# Define path to blacklist 
blacklist=/home/ext-smcclymont/SN4741_paper/mm10_ATAC_blacklist.bed

# Remove peak file header 
# Rearrange columns to bed format (chr start end name score strand), spoof strand column
  
# ! Check number of lines to remove from header and adjust code accordingly ! 
for file in *_peaks.xls;
	do
	awk 'NR > 26 { print }' < $file > temp
	awk -v OFS='\t' {'print $1,$2,$3,$10,$6,"+"'} temp > ${file%.xls}.bed
done

rm temp

# Count number of peaks called in each sample
wc -l *peaks.bed > peak_counts

# Remove the blacklisted regions from peaks and quantify number of peaks lost
# Use bedtools v2.30.0 to intersect peaks with blacklisted regions and exclude

for file in *peaks.bed; do
	bedtools intersect -v -wa -a $file -b $blacklist > ${file%.bed}_blacklisted.bed
done

wc -l *blacklisted* > blacklisted_counts

exit 0

```

Approx 70 peaks removed from each sample for overlapping blacklisted regions. Move blacklisted peaks to new folder (blacklisted) in parent dir for easy access. 

# Check libraries for quality: Observe in IGV

```{r, eval = F}

# Convert bedgraphs to TDFs for lower memory usage by IGV

# toTDF.sh

#!/bin/bash
#SBATCH --job-name=toTDF
#SBATCH --time=10:00:00

module load Java

for file in *.bdg;
	do
	 ~/Software/IGV_2.15.2/igvtools toTDF $file ${file%.bdg} mm10
done

exit 0

```

Move TDF files to new folder (toTDF) in parent directory. Transfer TDF and peak files to look at in IGV on laptop. Compare with SN4741 data. Pileups and their called peaks look reasonable to me. Low background. Visually, not particularly similar to the MB and FB samples. 

# Check libraries for quality: FRiP

The fraction of reads in peaks is a good metric of the background level of the libraries, which would indicate if the libraries are overdigested. Use DeepTools to calculate FRiP. DeepTools will produce one plot per library, but we will combine the library FRiPs and display on one plot.

```{r, eval = F}

# FRIP.sh

#!/bin/bash
#SBATCH --job-name=FRIP
#SBATCH --time=10:00:00
#SBATCH -N 1
#SBATCH -n 12

# Use deeptools v3.5.1 to plot enrichment of reads (bams) over the blacklisted peaks - once per lib
# Run from bams folder

module load libjpeg
module load bzip2

for file in *.bam; do
  plotEnrichment -b $file --BED ../blacklisted/${file%.bam}_peaks_blacklisted.bed -p 12 --outRawCounts ${file%.bam}.counts -v
done

exit 0

# Take the counts files and plot the FRiP on one graph.

# Combine the FRIP counts into one file
less *.counts > temp

# It looks ugly because it kept the header from each file. Remove with awk (by keeping only even numbered rows)
awk 'NR %2==0' temp > FRIP.summary

# Add back in header to help with R so it's nicer looking
sed -i "1i $(head -n 1 temp)" FRIP.summary
rm temp

# Read into R and plot
module load r
R #4.0.2

library(ggplot2)

# Read in
FRIP <- read.table("FRIP.summary", header = T)

# Bar plot
pdf(file = "SN4741_FRIP.pdf")
p <- ggplot(data=FRIP, aes(x=file, y=percent)) +
  geom_bar(stat="identity") + 
  coord_flip() + 
  theme_classic()
p
dev.off()

```

Move FRIP data to new folder (FRIP) in parent directory. Again 39-4 is the odd library out. Poor FRiP compared to the other libraries. A problem child is emerging. 

# Check libraries for quality: Fragment length distributions

Calculate the average fragment length from the mapping distance flag in the bam files to see if you have the characteristic nucleosome laddering expected in ATAC-seq libraries.

```{r, eval = F}

# fragment_length.sh

#!/bin/bash
#SBATCH --job-name=fragment_length
#SBATCH --time=2:00:00

module load samtools

# Use samtools and a horrible awk script to extract the mapping distance flag from the bams 
# Run from BAM folder

for file in *.bam;
	do
	samtools view $file | awk '$9>0' | cut -f 9 | sort | uniq -c | sort -b -k2,2n | sed -e 's/^[ \t]*//' > ${file%.bam}.fragment_length_count.txt
done

exit 0

# Prep the files for plotting in R - add column for sample name and grouping (37 or 39)
for file in *count.txt; do 
  sample=${file%.fragment_length_count.txt} # Keeps sample name
  grouping=${file:0:2} # Keeps first two characters, which will be "37" and "39" 
  awk -v OFS='\t' -v a=$sample -v b=$grouping {'print $0,a,b'} $file > ${sample}.fragment.R # Adds the columns
  done
  
# Combine like an rbind in R so have just one file to read into R 
# Combine 37 and 39 individually and also together
# cat doesn't take wildcards, so explicitly call files
  cat 37-1_SN4741_S1.fragment.R 37-2_SN4741_S3.fragment.R 37-3_SN4741_S2.fragment.R 37-4_SN4741_S8.fragment.R > combined.fragment.37.R
  cat 39-1_SN4741_S4.fragment.R 39-2_SN4741_S7.fragment.R 39-3_SN4741_S5.fragment.R 39-4_SN4741_S6.fragment.R> combined.fragment.39.R
  cat combined.fragment.37.R combined.fragment.39.R > combined.fragment.all.R

# Plot in R 
R
library(ggplot2)

# Read in
fragment_37 <- read.table("combined.fragment.37.R")
fragment_39 <- read.table("combined.fragment.39.R")
fragment_all <- read.table("combined.fragment.all.R")
fragment_all$V4 <- as.factor(fragment_all$V4) # Stops R from recognizing the temperatures as a continuous variable for plotting

# Plot
pdf("Fragment_length_37_39.pdf")
ggplot(fragment_37, aes(x = V2, y = V1, colour = V3)) + geom_line(linewidth = 1.05) + xlim(0,1000) + theme_classic()
ggplot(fragment_39, aes(x = V2, y = V1, colour = V3)) + geom_line(linewidth = 1.05) + xlim(0,1000) + theme_classic()
ggplot(fragment_all, aes(x = V2, y = V1, colour = V3, linetype = V4)) + geom_line(linewidth = 1.05) + xlim(0,1000) + theme_classic()
dev.off()

```

Move fragment length data and plots to new folder (fragment_lengths) in parent directory. See characteristic nucleosome laddering + the ~10bp sawtooth pattern of the DNA pitch. 

# Check libraries for biological relevance: Enrichment of reads at promoters

Promoters are generally open. We should expect to see enrichment of reads over top of promoters. Use DeepTools to check the pileup of reads over top of promoters on a genome wide basis.

## Download mm10 promoters 

Defining a promter as a transcriptional start site plus or minus 1000 basepairs. Use UCSC table browser to get the TSSs and then create a bed file of the surrounding 1000 bp.

```{r, eval = F}

# UCSC table browser: mm10, genes and gene predictions, NCBI RefSeq, refGene table, whole genome, selected fields (chrom, strand, txStart, txEnd, name2)
# Save as: mm10_RefSeq_txstart_txend

# Use R to rearrange columns, when on the negative strand to use the txend as the TSS, and remove duplicate TSSs
mm10_TSS <- read.table("mm10_RefSeq_txstart_txend")
colnames(mm10_TSS) <- c("chr", "strand", "start", "end", "geneID")
mm10_TSS$TSS <- ifelse(mm10_TSS$strand == "+", mm10_TSS$start, mm10_TSS$end) # If positive strand, TSS is the start site, otherwise its the end column
mm10_TSS$TSSplus <- mm10_TSS$TSS + 1000
mm10_TSS$TSSminus <- mm10_TSS$TSS - 1000
mm10_TSS$promBED <- paste0(mm10_TSS$chr, ":", mm10_TSS$TSSminus, "-", mm10_TSS$TSSplus) # Create a column of the chr:start-end so can deduplicate identical start sites
mm10_TSS <- mm10_TSS[!duplicated(mm10_TSS$promBED),] # deduplicate

mm10_proms <- mm10_TSS[,c(1,8,7,5,6,2) ] # rearrange to bed format, spoof score with TSS
write.table(x = mm10_proms, "mm10_promoters_1000bp.bed", quote = F, row.names = F, col.names = F, sep = "\t")

```

## Enrichment of reads at promoters 

```{r, eval = F}

# promoter_enrichment.sh

#!/bin/bash
#SBATCH --job-name=enrich_proms
#SBATCH --time=15:00:00
#SBATCH -N 1
#SBATCH -n 12

# Use deeptools v3.5.1 to plot enrichment of reads (bams) over the promoters
# Run from bams folder

module load libjpeg
module load bzip2

promoters=/home/ext-smcclymont/SN4741_paper/mm10_promoters_1000bp.bed

for file in *.bam; do
  sample=${file:0:4} # Just keeps the sample name with none of the extra ("SN4741_S#) that has been carrying through; keep it clean
	bamCoverage -p 12 -b $file -o ${sample}.bw
	computeMatrix reference-point -p 12 --referencePoint center -b 1000 -a 1000 -bs 50 -S ${sample}.bw -R $promoters -out ${sample}.matrix
	plotHeatmap --matrixFile ${sample}.matrix --outFileName ${sample}.heatmap.pdf --heatmapHeight 5 --heatmapWidth 5 
done

exit 0

```

All libraries look good - all have nice enrichment of reads over top of promoters. Have this weird blip downstream of TSS but it's present in all samples, so not too worried. Move files to new directory (promoter_enrichment) in parent directory.  
 