---
title: "ATAC-seq analysis of SN4741 cells"
author: "Sarah McClymont"
date: "November 11, 2022"
output: pdf_document
---

# Intro 

Generated ATAC-seq in the SN4741 cell line at both the permissive and restrictive temperatures (n=4/temperature). Want to see how the chromatin differs between the two states. Also want to consider how well they recapitulate the ATAC-seq data from primary E15.5 mouse dopaminergic neurons from both the forebrain and midbrain, which are presumably the cell type this cell line should be modelling.

# Pre-alignment QC

Use FastQC to check library quality. Summarize FastQC reports with MultiQC and check the generated HTML for sequencing quality issues.

```{r, eval = F}

# batch script: fastqc.sh

#!/bin/bash
#SBATCH --job-name=fastqc
#SBATCH --time=8:00:00
#SBATCH -N 1
#SBATCH -n 10

module load fastqc 

## fastqc v0.11.9
fastqc -t 10 *.fastq.gz

# Generate report with MultiQC
## pip3 install --user multiqc #v1.13
multiqc .

# Clean up files
mkdir FastQC
mv *fastqc* FastQC

exit 0

```

MultiQC report moved to Output/multiqc_report_fastqc.html

**Conclusions:** Look good in general. Sample 39-4 looks off in its GC content compared to other samples. High rates of duplication in 37-1 and 37-4 but the pool was apparently unbalanced and we sequenced these two very deeply so perhaps hit saturation. Adaptor contamination for all samples is functionally zero - approx 0.25% of reads end with the beginning of an adaptor - not worth trimming and shouldn't affect mapping at all. As always, all samples have sequence composition bias at the start of reads, as per usual for ATAC. Nothing of concern here - move forward with mapping!


# Align and call peaks

Use Bowtie2 to align to genome.  
UCSC mm10 indexes are the same as those used in the pcHiC. See NBB's code for source.  
  Note: NBB's genome files don't contain chrM so can't quantify %mito. Previously aligned these sequences and saw less than 1% contamination in all samples.

Deduplicate with Picard Tools. 
Index with SAMtools. 
Call peaks with MACS3.

Summarize reports with MultiQC.   

```{r, eval = F}

# alignment.sh

#!/bin/bash
#SBATCH --job-name=alignment
#SBATCH --time=30:00:00
#SBATCH -N 1
#SBATCH -n 12

# Load modules
  module load bowtie2 #v2.4.1
  module load samtools #v1.15.1
  module load picard #v2.26.11

# Define path to genome indices
  genome_index=/data/amccall2/work-zfs/NBB/SN4741-pcHiC/bowtie2_index/mm10

for file in *L001_R1_001.fastq.gz; do

# Make life easier by making the file names simpler and into an extensionlesss variable
  sample=${file%_L001_R1_001.fastq.gz}

# Write out sample name so can track progress/errors in slurm.out file
  echo $sample

# Align with Bowtie2
  # Comma separate input files, no white space between
  # Local alignment to softclip the ends
  # Increase pair distance to 1000 to avoid the 500bp blip on fragment length plot
  # Write out alignment stats (stderr) to file for multiqc summary
  bowtie2 -p 12 --local -X 1000 -x $genome_index -1 $file,${sample}_L002_R1_001.fastq.gz -2 ${sample}_L001_R2_001.fastq.gz,${sample}_L002_R2_001.fastq.gz -S ${sample}.sam 2>${sample}.align.stats

# Convert to bam and sort with samtools
  # view -bu turns into uncompressed bam for easy piping into sort
  samtools view -bu ${sample}.sam | samtools sort -o ${sample}_sorted.bam

# Deduplicate with picard tools
  picard MarkDuplicates I=${sample}_sorted.bam O=${sample}.bam M=${sample}_markdup.stats

# Index with samtools
  samtools index ${sample}.bam

# Remove intermediate files
	${sample}.sam
	${sample}_sorted.bam

# Call peaks! macs3 v3.0.0a7
	macs3 callpeak --nomodel -B -f BAMPE -t ${sample}.bam -n $sample --nolambda --gsize mm

	done

exit 0
  
# Compile alignment statistics with multiqc
	multiqc . --ignore "FastQC/"
	
```

**Conclusions:** Libraries look great! High alignment rate (~98% for all eight samples), and average duplication rate (â‰¤20%). Two/three samples have high duplication rate, but those are also the deepest sequenced, so seems like we are hitting saturation for them. 

Arrange outputs into folders: alignment, bams, FastQC, fastqs, and peaks for easier navigation.

# Removing blacklisted regions

*Source of blacklists:*
ENCODE blacklist (requires citation to ENCODE): https://sites.google.com/site/anshulkundaje/projects/blacklists
ATAC-seq specific blacklists (hg19, mm9, mm10; must be member of ATAC-seq forum to view): https://sites.google.com/site/atacseqpublic/atac-seq-analysis-methods/mitochondrialblacklists-1

Concatenated the two mm10 blacklists together; remove peaks that overlap these regions.

```{r, eval = F}

# blacklisted.sh

#!/bin/bash
#SBATCH --job-name=blacklisted
#SBATCH --time=1:00:00

# Load modules
module load bedtools

# Define path to blacklist 
blacklist=/home/ext-smcclymont/SN4741_paper/mm10_ATAC_blacklist.bed

# Remove peak file header 
# Rearrange columns to bed format (chr start end name score strand), spoof strand column
  
# ! Check number of lines to remove from header and adjust code accordingly ! 
for file in *_peaks.xls;
	do
	awk 'NR > 26 { print }' < $file > temp
	awk -v OFS='\t' {'print $1,$2,$3,$10,$6,"+"'} temp > ${file%.xls}.bed
done

rm temp

# Count number of peaks called in each sample
wc -l *peaks.bed > peak_counts

# Remove the blacklisted regions from peaks and quantify number of peaks lost
# Use bedtools v2.30.0 to intersect peaks with blacklisted regions and exclude

for file in *peaks.bed; do
	bedtools intersect -v -wa -a $file -b $blacklist > ${file%.bed}_blacklisted.bed
done

wc -l *blacklisted* > blacklisted_counts

exit 0

```

Approx 70 peaks removed from each sample for overlapping blacklisted regions. Move blacklisted peaks to new folder (blacklisted) in parent dir for easy access. 

# Check libraries for quality: Observe in IGV

```{r, eval = F}

# Convert bedgraphs to TDFs for lower memory usage by IGV

# toTDF.sh

#!/bin/bash
#SBATCH --job-name=toTDF
#SBATCH --time=10:00:00

module load Java

for file in *.bdg;
	do
	 ~/Software/IGV_2.15.2/igvtools toTDF $file ${file%.bdg} mm10
done

exit 0

```

Move TDF files to new folder (toTDF) in parent directory. Transfer TDF and peak files to look at in IGV on laptop. Compare with SN4741 data. Pileups and their called peaks look reasonable to me. Low background. Visually, not particularly similar to the MB and FB samples. 

