---
title: "Analysing E15.5 MB and FB DA neuron ATAC"
author: "Sarah McClymont"
date: "November 11, 2022"
output: pdf_document
---

# Intro 

Previously generated ATAC-seq data from E15.5 mouse dopaminergic neurons from the midbrain and forebrain (McClymont et al 2018). For this paper, we want to perform a comparison with the SN4741 data to see how well the cell line matches primary data. To do so, we want to make sure the original data is processed in the same way as the SN4741 for a fair comparison - reanalyse in parallel.

# Pre-alignment QC

Use FastQC to check library quality. Summarize FastQC reports with MultiQC and check the generated HTML for sequencing quality issues.

```{r, eval = F}

# batch script: fastqc_MBFB.sh

#!/bin/bash
#SBATCH --job-name=fastqc
#SBATCH --time=8:00:00
#SBATCH -N 1
#SBATCH -n 10

module load fastqc 

## fastqc v0.11.9
fastqc -t 10 *.fastq.gz

# Generate report with MultiQC
## pip3 install --user multiqc #v1.13
multiqc .

# Clean up files
mkdir FastQC
mv *fastqc* FastQC

exit 0

```

# Align and call peaks

Use Bowtie2 to align to genome.  
UCSC mm10 indexes are the same as those used in the pcHiC. See NBB's code for source.  

Deduplicate with Picard Tools. 
Index with SAMtools. 
Call peaks with MACS3.

Summarize reports with MultiQC.   

```{r, eval = F}

# alignment_MB.sh

#!/bin/bash
#SBATCH --job-name=alignment
#SBATCH --time=30:00:00
#SBATCH -N 1
#SBATCH -n 12

# Load modules
  module load bowtie2 #v2.4.1
  module load samtools #v1.15.1
  module load picard #v2.26.11

# Define path to genome indices
  genome_index=/data/amccall2/work-zfs/NBB/SN4741-pcHiC/bowtie2_index/mm10

#Hard to make for loop when some samples have been sequenced twice but most just once. Exclude them from the loop by temporarily moving to a new dir; analyse separately

for file in *R1.fastq.gz; do

# Make life easier by making the file names simpler and into an extensionlesss variable
  sample=${file%_R1.fastq.gz}

# Write out sample name so can track progress/errors in slurm.out file
  echo $sample

# Align with Bowtie2
  # Local alignment to softclip the ends
  # Increase pair distance to 1000 to avoid the 500bp blip on fragment length plot
  # Write out alignment stats (stderr) to file for multiqc summary
  bowtie2 -p 12 --local -X 1000 -x $genome_index -1 $file -2 ${sample}_R2.fastq.gz -S ${sample}.sam 2>${sample}.align.stats

# Convert to bam and sort with samtools
  # view -bu turns into uncompressed bam for easy piping into sort
  samtools view -bu ${sample}.sam | samtools sort -o ${sample}_sorted.bam

# Deduplicate with picard tools
  picard MarkDuplicates I=${sample}_sorted.bam O=${sample}.bam M=${sample}_markdup.stats

# Index with samtools
  samtools index ${sample}.bam

# Remove intermediate files
	rm ${sample}.sam
	rm ${sample}_sorted.bam

# Call peaks! macs3 v3.0.0a7
	macs3 callpeak --nomodel -B -f BAMPE -t ${sample}.bam -n $sample --nolambda --gsize mm

	done

exit 0

```

## Modifying loop for the resequenced samples 

```{r, eval = F}

# alignment_reseq.sh

#!/bin/bash
#SBATCH --job-name=alignment
#SBATCH --time=30:00:00
#SBATCH -N 1
#SBATCH -n 12

# Load modules
  module load bowtie2 #v2.4.1
  module load samtools #v1.15.1
  module load picard #v2.26.11

# Define path to genome indices
  genome_index=/data/amccall2/work-zfs/NBB/SN4741-pcHiC/bowtie2_index/mm10

# Align individually
  bowtie2 -p 10 --local -X 1000 -x $genome_index -1 FB2_ATAC_R1.fastq.gz,FB2_reseq_ATAC_R1.fastq.gz -2 FB2_ATAC_R2.fastq.gz,FB2_reseq_ATAC_R2.fastq.gz -S FB2_ATAC.sam 2>FB2.align.stats
  bowtie2 -p 10 --local -X 1000 -x $genome_index -1 MB3_ATAC_R1.fastq.gz,MB3_reseq_ATAC_R1.fastq.gz -2 MB3_ATAC_R2.fastq.gz,MB3_reseq_ATAC_R2.fastq.gz -S MB3_ATAC.sam 2>MB3.align.stats

# Reinitiate loop
for file in *.sam; do

# Make life easier by making the file names simpler and into an extensionlesss variable
  sample=${file%.sam}

# Write out sample name so can track progress/errors in slurm.out file
  echo $sample

# Convert to bam and sort with samtools
  # view -bu turns into uncompressed bam for easy piping into sort
  samtools view -bu $file | samtools sort -o ${sample}_sorted.bam

# Deduplicate with picard tools
  picard MarkDuplicates I=${sample}_sorted.bam O=${sample}.bam M=${sample}_markdup.stats
  
# Index with samtools
  samtools index ${sample}.bam
  
# Remove intermediate files
	rm ${sample}.sam
	rm ${sample}_sorted.bam

# Call peaks!
	macs3 callpeak --nomodel -B -f BAMPE -t ${sample}.bam -n $sample --nolambda --gsize mm

	done

exit 0
  
```

## Compile mapping statistics

Move the outputs back to the same folder and compile with MultiQC

```{r, eval = F}
# Compile alignment statistics with multiqc
	multiqc . --ignore "FastQC/" 
```

Arrange outputs into folders: alignment, bams, FastQC, fastqs, and peaks for easier navigation.


# Blacklisted peaks

*Source of blacklists:*
ENCODE blacklist (requires citation to ENCODE): https://sites.google.com/site/anshulkundaje/projects/blacklists
ATAC-seq specific blacklists (hg19, mm9, mm10; must be member of ATAC-seq forum to view): https://sites.google.com/site/atacseqpublic/atac-seq-analysis-methods/mitochondrialblacklists-1

Concatenated the two mm10 blacklists together; remove peaks that overlap these regions.

```{r, eval = F}

# blacklisted.sh

#!/bin/bash
#SBATCH --job-name=blacklisted
#SBATCH --time=1:00:00

# Load modules
module load bedtools

# Define path to blacklist 
blacklist=/home/ext-smcclymont/SN4741_paper/mm10_ATAC_blacklist.bed

# Remove peak file header 
# Rearrange columns to bed format (chr start end name score strand), spoof strand column
  
# ! Check number of lines to remove from header and adjust code accordingly ! 
for file in *_peaks.xls;
	do
	awk 'NR > 26 { print }' < $file > temp
	awk -v OFS='\t' {'print $1,$2,$3,$10,$6,"+"'} temp > ${file%.xls}.bed
done

rm temp

# Count number of peaks called in each sample
wc -l *peaks.bed > peak_counts

# Remove the blacklisted regions from peaks and quantify number of peaks lost
# Use bedtools v2.30.0 to intersect peaks with blacklisted regions and exclude

for file in *peaks.bed; do
	bedtools intersect -v -wa -a $file -b $blacklist > ${file%.bed}_blacklisted.bed
done

wc -l *blacklisted* > blacklisted_counts

exit 0

```

Approx 70 peaks removed from each sample for overlapping blacklisted regions. Move blacklisted peaks to new folder (blacklisted) in parent dir for easy access. 

# Check libraries for quality: Observe in IGV

```{r, eval = F}

# Convert bedgraphs to TDFs for lower memory usage by IGV

# toTDF.sh

#!/bin/bash
#SBATCH --job-name=toTDF
#SBATCH --time=10:00:00

module load Java

for file in *.bdg;
	do
	 ~/Software/IGV_2.15.2/igvtools toTDF $file ${file%.bdg} mm10
done

exit 0

```

Move TDF files to new folder (toTDF) in parent directory. Transfer TDF and peak files to look at in IGV on laptop. Compare with SN4741 data. 

# Check libraries for quality: FRiP

The fraction of reads in peaks is a good metric of the background level of the libraries, which would indicate if the libraries are overdigested. Use DeepTools to calculate FRiP. DeepTools will produce one plot per library, but we will combine the library FRiPs and display on one plot.

```{r, eval = F}

# FRIP.sh

#!/bin/bash
#SBATCH --job-name=FRIP
#SBATCH --time=10:00:00
#SBATCH -N 1
#SBATCH -n 12

# Use deeptools v3.5.1 to plot enrichment of reads (bams) over the blacklisted peaks - once per lib
# Run from bams folder

module load libjpeg
module load bzip2

for file in *.bam; do
  plotEnrichment -b $file --BED ../blacklisted/${file%.bam}_peaks_blacklisted.bed -p 12 --outRawCounts ${file%.bam}.counts -v
done

exit 0

# Take the counts files and plot the FRiP on one graph.

# Combine the FRIP counts into one file
less *.counts > temp

# It looks ugly because it kept the header from each file. Remove with awk (by keeping only even numbered rows)
awk 'NR %2==0' temp > FRIP.summary

# Add back in header to help with R so it's nicer looking
sed -i "1i $(head -n 1 temp)" FRIP.summary
rm temp

# Read into R and plot
module load r
R #4.0.2

library(ggplot2)

# Read in
FRIP <- read.table("FRIP.summary", header = T)

# Bar plot
pdf(file = "MBFB_FRIP.pdf")
p <- ggplot(data=FRIP, aes(x=file, y=percent)) +
  geom_bar(stat="identity") + 
  coord_flip() + 
  theme_classic()
p
dev.off()

```

Move FRIP data to new folder (FRIP) in parent directory. 

# Check libraries for quality: Fragment length distributions

Calculate the average fragment length from the mapping distance flag in the bam files to see if you have the characteristic nucleosome laddering expected in ATAC-seq libraries.

```{r, eval = F}

# fragment_length.sh

#!/bin/bash
#SBATCH --job-name=fragment_length
#SBATCH --time=2:00:00

module load samtools

# Use samtools and a horrible awk script to extract the mapping distance flag from the bams 
# Run from BAM folder

for file in *.bam;
	do
	samtools view $file | awk '$9>0' | cut -f 9 | sort | uniq -c | sort -b -k2,2n | sed -e 's/^[ \t]*//' > ${file%.bam}.fragment_length_count.txt
done

exit 0

# Prep the files for plotting in R - add column for sample name and grouping (MB and FB)
for file in *count.txt; do 
  sample=${file%_ATAC.fragment_length_count.txt} # Keeps sample name
  grouping=${file:0:2} # Keeps first two characters, which will be "MB" and "FB" 
  awk -v OFS='\t' -v a=$sample -v b=$grouping {'print $0,a,b'} $file > ${sample}.fragment.R # Adds the columns
  done
  
# Combine like an rbind in R so have just one file to read into R 
  cat FB1.fragment.R FB2.fragment.R FB3.fragment.R MB1.fragment.R MB2.fragment.R MB3.fragment.R > combined.fragment.R
  
# Plot in R 
R
library(ggplot2)

# Read in
combined <- read.table("combined.fragment.R")

pdf("Fragment_size_MB.pdf")
ggplot(combined, aes(x = V2, y = V1, colour = V3, linetype = V4)) + geom_line(linewidth = 1.05) + xlim(0,1000) + theme_classic()
dev.off()

```

Move fragment length data and plots to new folder (fragment_lengths) in parent directory. See characteristic nucleosome laddering + the ~10bp sawtooth pattern of the DNA pitch. 